README.md
# PostCSS Fontify [![npm version](https://badge.fury.io/js/postcss-fontify.svg)](https://badge.fury.io/js/postcss-fontify)

[PostCSS](https://github.com/postcss/postcss) plugin for generating font-face declarations from font files in a directory.

## Usage

Only font files with extensions `woff2`, `woff`, `ttf`, and `otf` will be recognized and used by the plugin. Any other files with different extensions in the same directory will not be considered. Additionally, the files in the directory must be named in accordance with a specific convention.

### Filename convention

The filename should have two parts. The first part should consist of the font name or font names written without any spaces and with the first letter of each word capitalized. The second part should be separated by a hyphen sign (-) and should contain the font weight and font style. The font style should only be included when the font is italic. The font weight can be one of the following: `Thin`, `ExtraLight`, `Light`, `Regular`, `Medium`, `SemiBold`, `Bold`, `ExtraBold` and `Black`. The two parts should be separated again by a capital letter. For example: `PragueSpecial-SemiBoldItalic.woff2.`

### Variable fonts

Fontify supports Variable Fonts, a new font technology that enables designers and developers to create a single font file with multiple variations of the font, including different weights, styles, and other properties. This technology allows for a more efficient and flexible way of delivering fonts to websites and other digital applications, reducing the file size and allowing for a more customized user experience.

The Variable Fonts page on the Adobe Type team's official website provides a more detailed explanation and examples: [https://typekit.com/variablefonts](https://typekit.com/variablefonts).

To correctly define the @font-face, Variable Fonts are only supported in the `woff2` format and require additional information about the `font-weight` range, which is expressed as number values. The name convention for Variable Fonts requires the keyword 'Var' followed by the font-weight start and end value, separated by an underscore.

If you need to determine which weights your Variable Fonts support or convert Variable font `ttf` to `woff2`, here are some helpful links:

- [Axis Praxis](https://www.axis-praxis.org/) A website for playing with OpenType Variable Fonts
- [Wakamai Fondue](https://wakamaifondue.com/) A tool for revealing all font information
- [V-Fonts](https://v-fonts.com/) A simple resource for finding and trying variable fonts
- [Every Things Fonts](https://everythingfonts.com/ttf-to-woff2)A font converter and compression tool
- [Woff2 Converter](https://github.com/google/woff2) A command line font compression tool from Google

## Input

From the font files in the specified folder:

```sh
`PragueSpecial-SemiBoldItalic.woff2.`
├── public
│   ├── fonts
│   │   ├── PragueSpecial-SemiBoldItalic.woff2
│   │   ├── PragueSpecial-SemiBoldItalic.woff
│   │   ├── PragueSpecial-SemiBoldItalic.ttf
│   │   ├── PragueSpecial-SemiBold.woff2
│   │   ├── PragueSpecial-SemiBold.woff
│   │   ├── PragueVariable-Var100_900.woff2
```

## Output

The following `@font-face` declaration is created:

```css
@font-face {
  family-name: "Prague Special";
  src: local("Prague Special"),
       url("/fonts/PragueSpecial-SemiBoldItalic.woff2") format("woff2"),
       url("/fonts/PragueSpecial-SemiBoldItalic.woff") format("woff"),
       url("/fonts/PragueSpecial-SemiBoldItalic.ttf") format("ttf");
  font-style: italic;
  font-weight: 600;
  font-display: swap;
}

@font-face {
  family-name: "Prague Special";
  src: local("Prague Special"),
       url("/fonts/PragueSpecial-SemiBold.woff2") format("woff2"),
       url("/fonts/PragueSpecial-SemiBold.woff") format("woff");
  font-style: normal;
  font-weight: 600;
  font-display: swap;
}

@font-face {
  family-name: "Prague Variable";
  src: local("Prague Variable"),
       url("/fonts/PragueVariable-SemiBold.woff2") format("woff2-variations");
  font-style: normal;
  font-weight: 100 900;
  font-display: swap;
}
```

The font folder can contain multiple font files with the same name and different formats. They will be all used in the `src` property.

## Install

**Step 1:** Install `postcss-fontify` plugin:

```sh
npm install --save-dev postcss postcss-fontify
```

**Step 2:** Check your project for existing PostCSS config: `postcss.config.js`
in the project root, `"postcss"` section in `package.json`
or `postcss` in bundle config.

**Step 3:** Add the plugin to plugin list:

```diff
module.exports = {
  plugins: [
+   require('postcss-font-face-generator')({
+     fontsDir: './public/fonts/',
+     fontPath: '/fonts/'
+   }),
    require('autoprefixer')
  ]
}
```

## Options

- `fontsDir`: The path to the directory containing the font files. Defaults to `'./fonts/'`.
- `fontPath`: The path for the `@font-face`.`url` declaration. Defaults to `'/fonts/'`.
- `local`: A boolean value indicating if a local font reference should be used. Defaults to `true`.
- `swap`: A boolean value indicating if the `font-display` property should be set to `'swap'`. Defaults to `true`.

[official docs]: https://github.com/postcss/postcss#usage


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2023-04-08

### Added

- Variable font support.

## [1.0.3] - 2023-04-06

### Added

- Change log.

### Changed

- Update the documentation.

### Removed

## [1.0.2] - 2023-04-05

### Fixed

- Rearrange the sequence of the source files based on the compression value of the font file.

## [1.0.1] - 2023-04-03

### Added

- Update package dependencies.

### Fixed

- Homepage url in package json file.

## [1.0.0] - 2023-03-31

### Added

- Basic functionality

- fix: correct src file order by file compression value


index.test.js
const postcss = require('postcss');
const { sortFiles } = require('./utils');
const plugin = require('./');


async function run (input, output, opts = {}) {
  const result = await postcss([plugin(opts)]).process(input, { from: undefined });
  expect(result.css).toEqual(output);
  expect(result.warnings()).toHaveLength(0);
}

describe('sortFiles', () => {
  it('should sort files based on weights', () => {
    const files = [
      { name: 'file1', srcFormat: 'ttf' },
      { name: 'file2', srcFormat: 'woff2' },
      { name: 'file3', srcFormat: 'woff' },
      { name: 'file4', srcFormat: 'otf' },
    ];

    const sortedFiles = files.sort(sortFiles);

    expect(sortedFiles).toEqual([
      { name: 'file2', srcFormat: 'woff2' },
      { name: 'file3', srcFormat: 'woff' },
      { name: 'file1', srcFormat: 'ttf' },
      { name: 'file4', srcFormat: 'otf' },
    ]);
  });

  it('should sort files with the same weight based on extension', () => {
    const files = [
      { name: 'file1', srcFormat: 'woff' },
      { name: 'file2', srcFormat: 'woff2' },
      { name: 'file3', srcFormat: 'otf' },
      { name: 'file4', srcFormat: 'ttf' },
      { name: 'file5', srcFormat: 'woff2' },
      { name: 'file6', srcFormat: 'ttf' },
      { name: 'file7', srcFormat: 'woff' },
      { name: 'file8', srcFormat: 'otf' },
    ];

    const sortedFiles = files.sort(sortFiles);

    expect(sortedFiles).toEqual([
      { name: 'file2', srcFormat: 'woff2' },
      { name: 'file5', srcFormat: 'woff2' },
      { name: 'file1', srcFormat: 'woff' },
      { name: 'file7', srcFormat: 'woff' },
      { name: 'file4', srcFormat: 'ttf' },
      { name: 'file6', srcFormat: 'ttf' },
      { name: 'file3', srcFormat: 'otf' },
      { name: 'file8', srcFormat: 'otf' },
    ]);
  });
});

describe('Generate font-face', () => {
it('The console should display a warning message when no font files are found in the given directory', async () => {
  const result = await postcss([plugin({ fontsDir: './fonts/empty/' })]).process('', { from: undefined });
  expect(result.warnings()).toHaveLength(1);
});

it('Create default output from files in font folder', async () => {
  const output = `@font-face {
    font-family: "Prague";
    src: local("Prague"), url("/fonts/Prague-Black.woff2") format("woff2"), url("/fonts/Prague-Black.woff") format("woff");
    font-weight: 900;
    font-style: normal;
    font-display: swap
}
@font-face {
    font-family: "Prague";
    src: local("Prague"), url("/fonts/Prague-Var.woff2") format("off2-variations");
    font-weight: 400;
    font-style: normal;
    font-display: swap
}
@font-face {
    font-family: "Prague";
    src: local("Prague"), url("/fonts/Prague-Var0_1000.woff2") format("off2-variations");
    font-weight: 0 1000;
    font-style: normal;
    font-display: swap
}
@font-face {
    font-family: "Prague";
    src: local("Prague"), url("/fonts/Prague-Var0_1000Italic.woff2") format("off2-variations");
    font-weight: 0 1000;
    font-style: italic;
    font-display: swap
}
@font-face {
    font-family: "Prague";
    src: local("Prague"), url("/fonts/Prague-Var300.woff2") format("off2-variations");
    font-weight: 300;
    font-style: normal;
    font-display: swap
}`;
  await run('', output, { fontsDir: './fonts/Prague/' });
});

it('Create output from files in font folder without swap option', async () => {
  const output = `@font-face {
    font-family: "Prague Special";
    src: local("Prague Special"), url("/fonts/PragueSpecial-ExtraLightItalic.woff2") format("woff2"), url("/fonts/PragueSpecial-ExtraLightItalic.woff") format("woff"), url("/fonts/PragueSpecial-ExtraLightItalic.ttf") format("ttf");
    font-weight: 200;
    font-style: italic
}`;
  await run('', output, { fontsDir: './fonts/PragueSpecial/', swap: false });
});

it('Create output from files in font folder without swap and local option', async () => {
  const output = `@font-face {
    font-family: "Prague Special";
    src: url("/fonts/PragueSpecial-ExtraLightItalic.woff2") format("woff2"), url("/fonts/PragueSpecial-ExtraLightItalic.woff") format("woff"), url("/fonts/PragueSpecial-ExtraLightItalic.ttf") format("ttf");
    font-weight: 200;
    font-style: italic
}`;
  await run('', output, { fontsDir: './fonts/PragueSpecial/', swap: false, local: false });
});
});


index.js
/**
 * @type {import('postcss').PluginCreator}
 */
const path = require('path');
const glob = require('glob');

const { sortFiles } = require('./utils');

module.exports = (opts = {}) => {
  const options = Object.assign({
    fontsDir: './fonts/',
    fontPath: '/fonts/',
    local: true,
    swap: true
  }, opts);

  const fontWeightMap = {
    Thin: '100',
    ExtraLight: '200',
    Light: '300',
    Regular: '400',
    Medium: '500',
    SemiBold: '600',
    Bold: '700',
    ExtraBold: '800',
    Black: '900'
  };

  return {
    postcssPlugin: 'postcss-fontify',

    Root (root, postcss) {
        const fontsDir = options.fontsDir;

        const fontFiles = glob.sync(`${fontsDir}/**/*.{woff2,woff,ttf,otf}`);
        if (fontFiles.length === 0) {
          root.warn(postcss.result, `[WARNING] No font files found in directory: ${fontsDir}`);
        }

        const fontFamilies = new Map();

        fontFiles.forEach(file => {
          const extension = path.extname(file);
          const fileName = path.basename(file, extension);
          const relativePath = path.relative(fontsDir, file);
          const srcPath = path.join(options.fontPath, relativePath);

          let [font, style] = fileName.split('-');
          const fontFamily = font.replace(/([a-z])([A-Z])/g, '$1 $2');
          let fontWeight = '400';
          let fontStyle = 'normal';
          let srcFormat = extension;

          if (style && style.includes('Italic')) {
            fontStyle = 'italic';
            style = style.replace('Italic', '');
          }

          if (style && style.includes('Var')) {
            const [start = Number.NaN, end = Number.NaN] = style.split('Var')[1].split('_').map(x => x === '' ? Number.NaN : Number(x));

            if (!Object.is(start, Number.NaN) && !Object.is(end, Number.NaN)) {
              fontWeight = `${start} ${end}`;
            }

            if (!Object.is(start, Number.NaN) && Object.is(end, Number.NaN)) {
              fontWeight = `${start}`;
            }

            srcFormat = 'woff2-variations';
          }

          if (style && fontWeightMap[style]) {
            fontWeight = fontWeightMap[style];
          }

          if (!fontFamilies.has(fileName)) {
            fontFamilies.set(fileName, {
              fontFamily,
              fontSrc: [{ srcFormat, srcPath }],
              fontWeight,
              fontStyle
            });
          } else {
            const font = fontFamilies.get(fileName);

            if (!font.fontSrc.some(obj => obj.srcFormat === srcFormat)) {
              font.fontSrc.push({ srcFormat, srcPath });
            }
            fontFamilies.set(fileName, font);
          }
        });

        fontFamilies.forEach((fontFace) => {
          let fontSrc = fontFace.fontSrc.sort(sortFiles).map((src) => `url("${src.srcPath}") format("${src.srcFormat.slice(1)}")`).join(', ').toString();

          if (options.local) {
            fontSrc = `local("${fontFace.fontFamily}"), ${fontSrc}`;
          }

          const nodes = [
            postcss.decl({
              prop: 'font-family',
              value: `"${fontFace.fontFamily}"`
            }),
            postcss.decl({
              prop: 'src',
              value: fontSrc
            }),
            postcss.decl({
              prop: 'font-weight',
              value: fontFace.fontWeight
            }),
            postcss.decl({
              prop: 'font-style',
              value: fontFace.fontStyle
            })
          ];

          if (options.swap) {
            nodes.push(
              postcss.decl({
                prop: 'font-display',
                value: 'swap',
              })
            );
          }

          const fontFaceDeclarations = postcss.root({ nodes });

          root.prepend(
            postcss.atRule({
              name: 'font-face',
              params: '',
              nodes: fontFaceDeclarations.nodes
            })
          );
        });

    }
  };
};

module.exports.postcss = true;
